### ISSUE TITLE
Feature shallow clone for Git brings error (Version 1.18.1)

### ISSUE THREAD

**Issue by climblinne** [conan-io/conan#5631]
To help us debug your issue please explain:

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've specified the Conan version, operating system version and any tool that can be relevant.
- [x] I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

Hello,

I am using Conan 1.18.1 on my Windows 10. 
I use internally `git describe` to build my version number of my package. This uses the latest tag, that has been put into git. 
Now with the new feature `shallow clone`, to safe time and space, I get the problem, that there are no `tags`in the history and I am unable to generate my package version.
It would be very nice, to have a property `shallow` in the `scm` tool. The best would be, to set the default, to  `shallow=False` to not break compatibility.

**Comment by jgsogo** [comment]
Hi, @climblinne. We didn't realize about this possible breaking behavior on https://github.com/conan-io/conan/pull/5514, let me label it as a `bug` and we may include the fix in `v1.18.2`.

Thanks for reporting it!

**Comment by lasote** [comment]
@climblinne could you provide an example of a recipe doing that? When does it fail? during a "conan export" command or during an install building the package in the cache? Are you using "auto" fields?

**Comment by climblinne** [comment]
The git repo must be tagged at a point in the history with e.g. `D0.0.1`.

In the `conanfile.py` I use the following:

```python
base = python_requires("base_library/0.1@demo/testing")
class pf_drv_imager_Conan(base.BaseLibrary):
    ...
    scm = base.BaseLibrary.get_git_scm_data()
    version = base.BaseLibrary.get_git_version()
```
And in my base class for all library the following:
``` python
    @staticmethod
    def get_git_version():
        try:
            scm = tools.Git()
            if not (Path(scm.folder) / '.git').exists():
                return None # No GIT folder in path
            # missing git describe
            describe = scm.run("describe --tags --long --dirty --match D[0-9]*.[0-9]*.[0-9]* --match RC[0-9]*.[0-9]*.[0-9]* --match V[0-9]*.[0-9]*.[0-9]*")
            e = re.match(r"(V|D|RC)([0-9]+)\.([0-9]+)\.([0-9]+)-([0-9]+)-g(.......)(-dirty)?", describe)
            
            if e and e.lastindex >= 6:
                if e.group(1) == "V":
                    branch_type = ""
                elif e.group(1) == "RC":
                    branch_type = "rc"
                elif e.group(1) == "D":
                    branch_type = "dev"
                else:
                    print("No matching name found in {branch_type}. Did you have at least one Tag 'D0.0.1'? ".format(branch_type=e.group(1)))
                version = "%s.%s.%s" % (e.group(2), e.group(3), e.group(4))
                add_commits = e.group(5)
                short_sha1 = e.group(6)
                if e.lastindex == 7:
                    dirty = "dirty."
                else:
                    dirty = ""
                
                branch = scm.get_branch().split(" ")[0] # Ignore text after whitespace (e.g. "HEAD (no branch)")
                version_text = "%s-%s.%s+%s%s.g%s" % (version, branch_type, add_commits, dirty, branch, short_sha1)
            else:
                print("No matching name found in git describe \"{describe}\". Did you have at least one Tag 'D0.0.1'? ".format(describe=describe))
            return version_text
        except Exception as ex:
            print("git describe failed (Did you have at least one Tag 'D0.0.1'?)")
            print(ex)
            return None

    @staticmethod
    def get_git_scm_data():
        return {
            "type": "git",
            "subfolder": ".",
            "url": "auto",
            "revision": "auto"}

```

I think you can drop the base class functionality and use the function straight from the main class.

**Comment by lasote** [comment]
I see, but once the recipe is in a server or in the cache, the "version" field shouldn't be used anymore,  the version is already frozen. As far as it doesn't raise an exception and I can see you are returning `None`, so it should work. 
Could you please paste here the error it is raising?
We probably anyway add an opt-out to the shallow clone but I need to understand why is it crashing to make sure we are not losing anything.

**Comment by lasote** [comment]
@climblinne ping

**Comment by climblinne** [comment]
Sorry, I was busy the last days. I have do reproduce it again.
As written in https://github.com/conan-io/conan/issues/5415 I use 'git describe' to generate the version number. Also I take the same 'git describe'  inside a CMake function to generate a 'version file header'.
Probably the problem comes from the CMake call of 'git describe'. 
Is there a possibility, to recognize, if Conan build inside '.conan' cache or just use 'conan install'. Then I could make a check dependent of the location. 
If inside '.conan' I just would check, that the repo is in non dirty state and that the git revision is identically. (Just to be safe, nobody changes something inside the folder)
When use another folder, there will be still the normal non-shallow clone and I could use the description as above.

**Comment by jgsogo** [comment]
Hi, @climblinne . We are proposing an opt-out for the shallow clone in #5677, it should be ready for `v1.18.2`. 

Nevertheless, have a look at the attibute `in_local_cache` ([docs](https://docs.conan.io/en/latest/reference/conanfile/attributes.html#in-local-cache)) it might be what you are looking for.

**Comment by climblinne** [comment]
Thanks @jgsogo . I think the `in_local_cache` could do the trick.

**Comment by lasote** [comment]
To be released in 1.18.2
